### ISSUE TITLE
x86 16bit CS segment override is wrong

### ISSUE THREAD

**Issue by simeonpilgrim** [NationalSecurityAgency/ghidra#2892]
I have x86 16bit DOS game engine I am reversing, and in one of the function in segment 0x28de:0827 it does a CALLF to 21e1:00d5 (which I had previously assumed was incorrectly label in that destination function)

```       
28de:0827 9a d5 00     CALLF    FUN_2ae1_00d5      undefined FUN_2ae1_00d5()
          e1 2a
								$U1210:2 = COPY CS
								SP = INT_SUB SP, 2:2
								$U1240:4 = CALLOTHER "segment", SS, SP
								STORE ram($U1240:4), $U1210:2
								CS = COPY 0x2ae1:2
								$U1210:2 = COPY 0x960c:2
								SP = INT_SUB SP, 2:2
								$U1240:4 = CALLOTHER "segment", SS, SP
								STORE ram($U1240:4), $U1210:2
								CALL *[ram]0x2aee5:4
```

In FUN_2ae1_00d5 it had an `assume CS = 0x28de` and the second instruction was a `MOV   word ptr CS:[0x58],ES` this was the part that made me think "something is wrong" the address was showing as fun_1f5e_0a74

the PCode for the MOV is:

```
2ae1:00d6 2e 8c 06        MOV   word ptr CS:[0x58],ES
          58 00
								$U570:4 = INT_RIGHT 0x2aeeb:4, 4:4
								$U590:4 = INT_AND $U570:4, 0xf000:4
								CS = SUBPIECE $U590:4, 0:4
								$U700:4 = CALLOTHER "segment", CS, 0x58:2
								$Uf10:2 = COPY ES
								STORE ram($U700:4), $Uf10:2

```
which is aligned with the code in the ia.sinc

```
currentCS: CS is protectedMode=0 & CS { tmp:4 = (inst_next >> 4) & 0xf000; CS = tmp:2; export CS; }
currentCS: CS is protectedMode=1 & CS { tmp:4 = (inst_next >> 16) & 0xffff; CS = tmp:2; export CS; }

seg16: currentCS: is segover=1 & currentCS	{ export currentCS; }
...

:MOV rm16,Sreg      is vexMode=0 & byte=0x8c; rm16 & Sreg ...               { rm16 = Sreg; }

```

So basically, while the function thinks CS is 0x28de, the CS is 0x2ae1, the disassembly decides to guess CS based on current address masked 0xF000, thus the code is looking at 0x2000:58 when it should be  0x2ae1:58

for example if the CS was 0x1230 and we were IP of 0x4560, thus at  0x16860 this above would decide CS was 0x1000 and we would be pointing at the address by 2300 bytes.
 
**Environment (please complete the following information):**
 - OS: Window 10
 - Java Version: 12.0.1
 - Ghidra Version: 9.2.1
 - Ghidra Origin: official ghidra-sre.org distro

**Comment by simeonpilgrim** [comment]
at this point are altered the CS seg16 for CS to be normal:

    seg16: CS: is segover=1 & CS	{ export CS; }

and select the whole function and corrected the incorrect assume on CS
and then dropped the incorrect memory references.

Would interesting to know a simple way to fix the last two globally/automatically verse line by line.

**Comment by tomsons26** [comment]
In 11.3.1 this issue behaves like this, initial analysis sets up CS overrides correctly but anywhere CS is used the resulting memory reference is wrong

![Image](https://github.com/user-attachments/assets/bd7556dd-f028-4e3c-9e95-9da1ff345833)
In this code the CS is correctly 0x39DA, first memory reference should be 39DA:0100

If i delete the memory reference and create it again
![Image](https://github.com/user-attachments/assets/e808efc6-44df-4f79-bab3-9e52412974d2)
the address in disassembly is correct, but the pseudocode insists using the wrong address still.

Overall seems ghidra can deduce the address correctly, but the decompiler insists on using the old one still.
All these issues seem to stem from currentCS or seg16 in sleigh, i tried modifying it but results were always inconsistent, don't really understand how either work.

**Comment by tomsons26** [comment]
On the upside, looks like calls are with correct addresses?

**Comment by Wall-AF** [comment]
Check the 2 function calls above either: 1) have the correct calling convention; or 2) adjust the stack depth manually.

From the image, the "stack depth" changes aren't visible.  To change this, click the "Edit the Listing fields" to show what's below
![Image](https://github.com/user-attachments/assets/f9a419b0-2212-46cf-bc65-d7be75e94a5c)
and right click the row containing "Mnemonic" and add "Stack Depth"
(Sorry if you already know this!)
